test/PidTest.cpp

```diff
@@ -0,0 +1,191 @@
+/**
+ * @file PidTest.cpp
+ * @brief PidGainクラス,Pidクラスをテストする
+ * @author TakumaChihaya
+ */
+
+#include "Pid.h"
+#include <gtest/gtest.h>
+
+namespace etrobocon2024_test {
+  TEST(PidGainTest, gain)
+  {
+    double expected_p = 0.1;
+    double expected_i = 0.03;
+    double expected_d = 0.612;
+    PidGain actualPidGain(expected_p, expected_i, expected_d);
+    EXPECT_DOUBLE_EQ(expected_p, actualPidGain.kp);
+    EXPECT_DOUBLE_EQ(expected_i, actualPidGain.ki);
+    EXPECT_DOUBLE_EQ(expected_d, actualPidGain.kd);
+  }
+
+  TEST(PidGainTest, gainZero)
+  {
+    double expected_p = 0.0;
+    double expected_i = 0.0;
+    double expected_d = 0.0;
+    PidGain actualPidGain(expected_p, expected_i, expected_d);
+    EXPECT_DOUBLE_EQ(expected_p, actualPidGain.kp);
+    EXPECT_DOUBLE_EQ(expected_i, actualPidGain.ki);
+    EXPECT_DOUBLE_EQ(expected_d, actualPidGain.kd);
+  }
+
+  TEST(PidGainTest, gainMinus)
+  {
+    double expected_p = -0.5;
+    double expected_i = -0.2;
+    double expected_d = -0.3;
+    PidGain actualPidGain(expected_p, expected_i, expected_d);
+    EXPECT_DOUBLE_EQ(expected_p, actualPidGain.kp);
+    EXPECT_DOUBLE_EQ(expected_i, actualPidGain.ki);
+    EXPECT_DOUBLE_EQ(expected_d, actualPidGain.kd);
+  }
+
+  TEST(PidTest, calculatePid)
+  {
+    constexpr double DELTA = 0.01;
+    double expected_p = 0.6;
+    double expected_i = 0.02;
+    double expected_d = 0.03;
+    double targetValue = 70;
+    Pid actualPid(expected_p, expected_i, expected_d, targetValue);
+    double currentValue = 20;
+    double preDeviation = 0;
+    double currentDeviation = (targetValue - currentValue);
+    double p = currentDeviation * expected_p;
+    double i = currentDeviation * DELTA * expected_i;
+    double d = (currentDeviation - preDeviation) * expected_d / DELTA;
+    double expected = p + i + d;
```

Comment:
```suggestion
    double expected = 実際に上記パラメータで算出される操作量;
```
みたいにしないと意味なくない？
現状、テスト対象のクラスと同じことを再度やってるだけな気がする。
`実際に上記パラメータで算出される操作量`の部分は手元で計算した値を入れないといけないんだけど、細かい数値がですぎて、無理かな？

Reply:
calculatepidが期待通りに動いてくれるかのテストなので，個人的には手元でしても同じ計算過程になるなら，このままでよいと思うのですが．．
理解できているか不安なので自分も他の方の意見も聞きたいです．

Reply:
テストを正しく実行するという意味では柿木さんの指摘のようにすべきだとは思います。
プログラムの処理が手元と同じ計算になるかどうかをチェックするのがテストで、それが正しい前提にするのはテストの意味が怪しくなる気もします。
前年になおせていなくて申し訳ないですが、コメントで計算過程を記して操作量を初期化して、その値と比較した方がいい派です。
他の方の意見お待ちしております🙇‍♂️

Reply:
確かに、手元の計算と同じになるのかのチェックがテストですね。手元で計算したものも作っておきます。

Reply:
テストが手元での計算結果でも通りました．
なのですが，やはり正しい計算過程で改めて計算したexpectedと，calculatePidで計算したactualを比較してcalculatePidが正しく機能しているかをテストするという，テストでの処理(?)は変わらないので手元で同じ計算をしても機械に任せてもどっちでも同じことをしていると思います．また，実際に本番で計算するのは機械ですし，改めて正しい計算過程を与えて計算された結果をexpectedとした方が正しくテストできていると言えると個人的に感じます．
あくまでもテスト経験0の個人的な意見なので，いくらでも異議受け付けます🙇‍♂️

Reply:
手計算のミスを減らすという点なら良いと思うけれど、
宅間君の考え方だと、calculatePidの実装が間違っていてもexpectedとactualが同じになるよね。
手元で同じ計算をしても機械に任せてもどっちでも同じになるっていうのは、結果論だと思う。
実装と手計算の考え方(計算過程)が同じでも、実装の部分で間違ってる可能性が万が一にもあると思うから、
僕の意見としては、木村君が述べたように、コメントで計算過程を記して操作量を初期化して、その値と比較した方がいいと思う。

Reply:
すみません、言葉不足でした。僕の考えではcalculatePidの処理は意識せず、改めて定義に戻って計算処理を記述して、その結果値をexpectedにするなら機械に計算させてもいいのではという考えでした。

Reply:
なるほど、手計算を機械に計算させるってことね。
それは良さそう！手計算のミス減らせるし、楽できるし良いね！

Reply:
[ネットのPID制御プログラム例](https://controlabo.com/pid-program/#toc2)での積分処理では，katlabのコードに置き換えると，
`integral += (currentDeviation + prevDeviation) * delta/2;`
となっていて，現在のkatlabの処理は
` integral += currentDeviation * delta;`
となっています．
実際に計算すると，0.004~0.005ほどの誤差が出たのですが，どうしますか？例えば，kp = 0.6，ki = 0.02，kd = 0.03の時，現在は"180.00999999999999"で，変更すると"180.005"になります．